### ISSUE TITLE
Simulating Dynamic Ambisonics Scenes on-the-fly

### ISSUE THREAD

**Issue by root-goksenin** [DavidDiazGuerra/gpuRIR#86]
Hey,

Great work! 

I am curious about the feasibility of on-the-fly dynamic ambisonics simulation. My scenes will contain a single source, and diffused/localized noise fields. I want the source to be dynamic, and the noise to be dynamic if it is localized. How realistic would it be use this library to fulfill my goal?

**Comment by DavidDiazGuerra** [comment]
Hi Göksenin,

Unfortunately, gpuRIR doesn't support ambisonics. Ambisonics are, with frequency-dependent absorptions, the most requested addition for the library and I have an idea about how to integrate both in an efficient way, but unfortunately I don't have time to implement it myself. If you think you could help with the implementation, or know someone who could and might be interested, I would be happy to supervise the work, but I should warn that it requires CUDA coding.

Best,
David

**Comment by root-goksenin** [comment]
Hey, 

That's fine for me if it does not support it! Unfortunately, I do not have much CUDA knowledge that I can contribute :/ 

However, if the library supports spherical microphones, then I can derive the ambisonics using ambisonics signal matching (ASM), albeit with some degradation compared to ground truth ambisonics.  I would still be very interested to try this out!

**Comment by DavidDiazGuerra** [comment]
Unfortunately, modeling the rigid body of a spherical array would require frequency-dependent directivities, so gpuRIR can only simulate "open" spherical arrays with omnidirectional microphones. If that's enough for you to derive the ambisonic signals, you can use gpuRIR placing the microphones on a spherical shape. The main problem will be that, to increase the accuracy of the ambisonic signals (especially if you need high orders), you will need to increase the number of microphones, which will increase the simulation time.

**Comment by root-goksenin** [comment]
Yes! That'd do it. I could get a tethrahedral mic layout and derive FOAs, as I won't work with HOAs for now. 

And could you maybe also comment on the feasibility of generating dynamic scenes on the fly?

**Comment by DavidDiazGuerra** [comment]
What do you mean by "on the fly"? You mean to be played in real time or during the training of a neural network?

**Comment by root-goksenin** [comment]
During the training of a neural network

**Comment by DavidDiazGuerra** [comment]
In that case, it should be possible, but it will most likely slow down your training. How much it slows will depend on the number of microphones, the number of trajectory points and sources, and how slow the training of the model already is (if every training step is very slow, the simulation time might be negligible, but this is not the case usually). 

In my [Cross3D](https://github.com/DavidDiazGuerra/Cross3D) and [icoDOA](https://github.com/DavidDiazGuerra/icoDOA) works, I trained neural networks for sound source localization generating the scenes on the fly during training. I was using 12 microphones and one source with a 156-point trajectory. Tbh, I'm not sure if I would do it like this again, but I didn't analyze the impact on the training time of doing the simulations on the fly compared with creating a fixed database in advance and then just loading the audio files during training.

**Comment by root-goksenin** [comment]
Thank you very much for your help! 

Now, my pre-training consists of only static sources. Once I test my model on dynamic downstream tasks, my model struggles with a linear-probe and thus a temporal model over the learnt embeddings is needed. I think it'll be negligible to simulate 10-20% of the scenes to be dynamic, and hopefully should boost the performance of the model on dynamic scenes. I have limited disk space and IO, so saving many RIRs to simulate trajectories, I think, is not admissible. I will give it a try and will let you know how it goes.
